src/main/java/de/tum/in/www1/artemis/service/CourseScoreCalculationService.java

```diff
@@ -0,0 +1,293 @@
+
+import java.time.ZonedDateTime;
+import java.util.*;
+
+import javax.annotation.Nullable;
+
+import org.springframework.stereotype.Service;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+
+import de.tum.in.www1.artemis.domain.*;
+import de.tum.in.www1.artemis.domain.enumeration.*;
+import de.tum.in.www1.artemis.domain.participation.Participation;
+import de.tum.in.www1.artemis.domain.participation.StudentParticipation;
+import de.tum.in.www1.artemis.domain.plagiarism.PlagiarismCase;
+import de.tum.in.www1.artemis.domain.plagiarism.PlagiarismVerdict;
+import de.tum.in.www1.artemis.repository.*;
+import de.tum.in.www1.artemis.repository.plagiarism.PlagiarismCaseRepository;
+import de.tum.in.www1.artemis.service.plagiarism.PlagiarismCaseService.PlagiarismMapping;
+import de.tum.in.www1.artemis.web.rest.dto.*;
+
+/**
+ * Service Implementation for calculating course scores.
+ * Adapted from the implementation at course-score-calculation.service.ts.
+ */
+@Service
+public class CourseScoreCalculationService {
+
+    private static final double SCORE_NORMALIZATION_VALUE = 0.01;
+
+    private final StudentParticipationRepository studentParticipationRepository;
+
+    private final ExerciseRepository exerciseRepository;
+
+    private final PlagiarismCaseRepository plagiarismCaseRepository;
+
+    public CourseScoreCalculationService(StudentParticipationRepository studentParticipationRepository, ExerciseRepository exerciseRepository,
+            PlagiarismCaseRepository plagiarismCaseRepository) {
+        this.studentParticipationRepository = studentParticipationRepository;
+        this.exerciseRepository = exerciseRepository;
+        this.plagiarismCaseRepository = plagiarismCaseRepository;
+    }
+
+    /**
+     * Calculates max and reachable max points for the given course and the student scores for the given student ids
+     * and takes the effects of related plagiarism verdicts on the grade into account. Implementation is adapted from course-score-calculation.service.ts.
+     * <p>
+     * If there is a single student id in studentIds, the student id will be filtered in the database as an optimization
+     * wherever possible.
+     *
+     * @param courseId   the id of the course to calculate
+     * @param studentIds the id of the students whose scores in the course will be calculated.
+     * @return the max and reachable max points for the given course and the student scores with related plagiarism verdicts for the given student ids
+     */
+    public CourseScoresDTO calculateCourseScores(long courseId, Collection<Long> studentIds) {
+        Set<Exercise> courseExercises = exerciseRepository.findAllExercisesByCourseId(courseId);
+        if (courseExercises.isEmpty()) {
+            return null;
+        }
+
+        double maxPointsInCourse = 0.0; // the sum of all included exercises, whose due date is over or unset or who are automatically assessed and assessment is done
```

Comment:
Maybe this could be more readable as streams?
Something like this.
```java
var maxPointsInCourse = courseExercises.stream()
  .filter(exercise -> includeIntoScoreCalculation(exercise))
  .filter(exercise -> exercise.getIncludedInOverallScore() == includedInOverallScore.INCLUDED_COMPLETELY)
  .map(exercise -> exercise.getMaxPoints())
  .sum();

var maxPointsReachableInExercise = courseExercises.stream()
  .filter(exercise -> includeIntoScoreCalculation(exercise))
  .filter(exercise -> exercise.getIncludedInOverallScore() == includedInOverallScore.INCLUDED_COMPLETELY)
  .filter(exercise -> isAssessmentDone(exercise))
  .map(exercise -> exercise.getMaxPoints()).sum();

```

Reply:
Maybe you could even refactor `includeIntoScoreCalculation()` to `Stream<Exercise> filterIncludedInScore(List<Exercise> exercises)`.

Reply:
In general I agree that using streams are more elegant, in this case I am hesitant to do it because of two reasons:

1. The filtering logic is almost always the same for these two point types so if we need to add a filter to the first stream, we would probably need to add it to the second one as well and is it easier to forget it to the second one when they are independent and the conditions are duplicated.
2. This method is adapted from the corresponding client implementation in [course-score-calculation.service.ts](https://github.com/ls1intum/Artemis/blob/develop/src/main/webapp/app/overview/course-score-calculation.service.ts#L36-L43) so if a developer would like to make changes in the calculation logic I think it may take less mental load to integrate it in both places when the code structure is similar.